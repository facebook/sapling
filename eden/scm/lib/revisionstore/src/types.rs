/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

use minibytes::Bytes;
#[cfg(any(test, feature = "for-tests"))]
use quickcheck_arbitrary_derive::Arbitrary;
use serde_derive::Deserialize;
use serde_derive::Serialize;
use types::Key;
use types::Sha256;

/// Kind of content hash stored in the LFS pointer. Adding new types is acceptable, re-ordering or
/// removal is forbidden.
#[derive(
    Serialize,
    Deserialize,
    PartialEq,
    Eq,
    Debug,
    Hash,
    Ord,
    PartialOrd,
    Clone
)]
#[cfg_attr(any(test, feature = "for-tests"), derive(Arbitrary))]
pub enum ContentHash {
    Sha256(#[serde(with = "types::serde_with::sha256::tuple")] Sha256),
}

#[derive(Clone, PartialEq, Eq, Debug, Hash, Ord, PartialOrd)]
#[cfg_attr(any(test, feature = "for-tests"), derive(Arbitrary))]
pub enum StoreKey {
    HgId(Key),
    /// The Key is a temporary workaround to being able to fallback from the LFS protocol to the
    /// non-LFS protocol. Do not depend on it as it will be removed.
    Content(ContentHash, Option<Key>),
}

impl ContentHash {
    pub fn sha256(data: &Bytes) -> Self {
        use sha2::Digest;

        let mut hash = sha2::Sha256::new();
        hash.update(data);

        let bytes: [u8; Sha256::len()] = hash.finalize().into();
        ContentHash::Sha256(Sha256::from(bytes))
    }

    pub fn unwrap_sha256(self) -> Sha256 {
        match self {
            ContentHash::Sha256(hash) => hash,
        }
    }

    pub(crate) fn sha256_ref(&self) -> &Sha256 {
        match self {
            ContentHash::Sha256(hash) => hash,
        }
    }
}

impl StoreKey {
    pub fn hgid(key: Key) -> Self {
        StoreKey::HgId(key)
    }

    pub fn content(hash: ContentHash) -> Self {
        StoreKey::Content(hash, None)
    }

    pub fn maybe_into_key(self) -> Option<Key> {
        match self {
            StoreKey::HgId(key) => Some(key),
            StoreKey::Content(_, k) => k,
        }
    }

    pub fn maybe_as_key(&self) -> Option<&Key> {
        match self {
            StoreKey::HgId(key) => Some(key),
            StoreKey::Content(_, k) => k.as_ref(),
        }
    }
}

impl From<Key> for StoreKey {
    fn from(k: Key) -> Self {
        StoreKey::HgId(k)
    }
}

impl<'a> From<&'a Key> for StoreKey {
    fn from(k: &'a Key) -> Self {
        StoreKey::HgId(k.clone())
    }
}

impl From<ContentHash> for StoreKey {
    fn from(hash: ContentHash) -> Self {
        StoreKey::Content(hash, None)
    }
}

impl<'a> From<&'a ContentHash> for StoreKey {
    fn from(hash: &'a ContentHash) -> Self {
        StoreKey::Content(hash.clone(), None)
    }
}
