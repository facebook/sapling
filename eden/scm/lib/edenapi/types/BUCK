load("@fbcode_macros//build_defs:rust_library.bzl", "rust_library")
load("@fbcode_macros//build_defs:rust_unittest.bzl", "rust_unittest")

oncall("sapling")

rust_library(
    name = "edenapi_types",
    srcs = glob(["src/**/*.rs"]),
    autocargo = {"cargo_toml_config": {
        "dependencies_override": {"dependencies": {"sapling-dag-types": {"features": ["for-tests"]}}},
        "lib": {"name": "edenapi_types"},
        "package": {
            "authors": ["Meta Source Control Team <sourcecontrol-dev@meta.com>"],
            "homepage": "https://sapling-scm.com/",
            "license": "MIT",
            "name": "sapling-edenapi_types",
            "repository": "https://github.com/facebook/sapling",
        },
    }},
    crate_root = "src/lib.rs",
    features = ["for-tests"],
    test_deps = [
        "fbsource//third-party/rust:quickcheck",
        "fbsource//third-party/rust:quickcheck_macros",
        "fbsource//third-party/rust:serde_json",
        "//common/rust/shed/quickcheck_arbitrary_derive:quickcheck_arbitrary_derive",
        "//eden/scm/lib/insta_ext:insta_ext",
    ],
    deps = [
        "fbsource//third-party/rust:anyhow",
        "fbsource//third-party/rust:bytes",
        "fbsource//third-party/rust:paste",
        "fbsource//third-party/rust:quickcheck",
        "fbsource//third-party/rust:serde",
        "fbsource//third-party/rust:serde_bytes",
        "fbsource//third-party/rust:serde_derive",
        "fbsource//third-party/rust:thiserror",
        ":type_macros",
        "//common/rust/shed/quickcheck_arbitrary_derive:quickcheck_arbitrary_derive",
        "//eden/scm/lib/blob:blob",
        "//eden/scm/lib/dag/dag-types:dag-types",
        "//eden/scm/lib/manifest-augmented-tree:manifest-augmented-tree",
        "//eden/scm/lib/minibytes:minibytes",
        "//eden/scm/lib/revisionstore:revisionstore_types",
        "//eden/scm/lib/types:types",
        "//eden/scm/lib/util/format-util:format-util",
    ],
)

rust_library(
    name = "type_macros",
    srcs = glob(["proc_macros/*.rs"]),
    autocargo = {
        "cargo_toml_config": {
            "lib": {"name": "type_macros"},
            "package": {
                "authors": ["Meta Source Control Team <sourcecontrol-dev@meta.com>"],
                "homepage": "https://sapling-scm.com/",
                "license": "MIT",
                "name": "sapling-type_macros",
                "repository": "https://github.com/facebook/sapling",
            },
        },
        "cargo_toml_dir": "proc_macros",
    },
    crate_root = "proc_macros/lib.rs",
    proc_macro = True,
    deps = [
        "fbsource//third-party/rust:proc-macro2",
        "fbsource//third-party/rust:quote",
        "fbsource//third-party/rust:syn1",
    ],
)

rust_unittest(
    name = "macro_test",
    srcs = ["test/basic_test.rs"],
    crate_root = "test/basic_test.rs",
    deps = [
        "fbsource//third-party/rust:quickcheck",
        "fbsource//third-party/rust:serde",
        "fbsource//third-party/rust:serde_json",
        ":edenapi_types",
        ":type_macros",
        "//common/rust/shed/quickcheck_arbitrary_derive:quickcheck_arbitrary_derive",
    ],
)
