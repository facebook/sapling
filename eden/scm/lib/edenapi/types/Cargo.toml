# @generated by autocargo from //eden/scm/lib/edenapi/types:[edenapi_types,macro_test]

[package]
name = "sapling-edenapi_types"
version = "0.1.0"
authors = ["Meta Source Control Team <sourcecontrol-dev@meta.com>"]
edition = "2021"
homepage = "https://sapling-scm.com/"
repository = "https://github.com/facebook/sapling"
license = "MIT"

[lib]
name = "edenapi_types"

[[test]]
name = "macro_test"
path = "test/basic_test.rs"

[dependencies]
anyhow = "1.0.95"
blob = { version = "0.1.0", path = "../../blob" }
bytes = { version = "1.9.0", features = ["serde"] }
manifest-augmented-tree = { version = "0.1.0", path = "../../manifest-augmented-tree" }
paste = "1.0.14"
quickcheck = "1.0"
quickcheck_arbitrary_derive = { version = "0.1.0", git = "https://github.com/facebookexperimental/rust-shed.git", branch = "main" }
sapling-dag-types = { version = "0.1.0", path = "../../dag/dag-types", features = ["for-tests"] }
sapling-format-util = { version = "0.1.0", path = "../../util/format-util" }
sapling-minibytes = { version = "0.1.0", path = "../../minibytes" }
sapling-revisionstore_types = { version = "0.1.0", path = "../../revisionstore/types" }
sapling-type_macros = { version = "0.1.0", path = "proc_macros" }
sapling-types = { version = "0.1.0", path = "../../types" }
serde = { version = "1.0.185", features = ["derive", "rc"] }
serde_bytes = "0.11"
serde_derive = "1.0.185"
thiserror = "2.0.12"

[dev-dependencies]
quickcheck_macros = "1.0"
sapling-insta_ext = { version = "0.1.0", path = "../../insta_ext" }
serde_json = { version = "1.0.140", features = ["float_roundtrip", "unbounded_depth"] }

[features]
default = ["for-tests"]
for-tests = []
