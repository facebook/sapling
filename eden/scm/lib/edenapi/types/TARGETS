load("@fbcode_macros//build_defs:rust_library.bzl", "rust_library")
load("@fbcode_macros//build_defs:rust_unittest.bzl", "rust_unittest")
load("@fbsource//tools/build_defs:glob_defs.bzl", "glob")

oncall("mercurial_release")

rust_library(
    name = "edenapi_types",
    srcs = glob(["src/**/*.rs"]),
    autocargo = {
        "cargo_toml_config": {
            "dependencies_override": {
                "dependencies": {
                    "dag-types": {"features": ["for-tests"]},
                },
            },
        },
    },
    crate_root = "src/lib.rs",
    features = ["for-tests"],
    test_deps = [
        "fbsource//third-party/rust:quickcheck",
        "fbsource//third-party/rust:quickcheck_macros",
        "fbsource//third-party/rust:serde_json",
        "//common/rust/shed/quickcheck_arbitrary_derive:quickcheck_arbitrary_derive",
        "//eden/scm/lib/insta_ext:insta_ext",
    ],
    deps = [
        "fbsource//third-party/rust:anyhow",
        "fbsource//third-party/rust:bytes",
        "fbsource//third-party/rust:paste",
        "fbsource//third-party/rust:quickcheck",
        "fbsource//third-party/rust:serde",
        "fbsource//third-party/rust:serde_bytes",
        "fbsource//third-party/rust:serde_derive",
        "fbsource//third-party/rust:thiserror",
        ":type_macros",
        "//common/rust/shed/quickcheck_arbitrary_derive:quickcheck_arbitrary_derive",
        "//eden/scm/lib/dag/dag-types:dag-types",
        "//eden/scm/lib/revisionstore:revisionstore_types",
        "//eden/scm/lib/types:types",
    ],
)

rust_library(
    name = "type_macros",
    srcs = glob(["proc_macros/*.rs"]),
    autocargo = {
        "cargo_toml_dir": "proc_macros",
    },
    crate_root = "proc_macros/lib.rs",
    proc_macro = True,
    deps = [
        "fbsource//third-party/rust:proc-macro2",
        "fbsource//third-party/rust:quote",
        "fbsource//third-party/rust:syn1",
    ],
)

rust_unittest(
    name = "macro_test",
    srcs = ["test/basic_test.rs"],
    crate_root = "test/basic_test.rs",
    deps = [
        "fbsource//third-party/rust:quickcheck",
        "fbsource//third-party/rust:serde",
        "fbsource//third-party/rust:serde_json",
        ":edenapi_types",
        ":type_macros",
        "//common/rust/shed/quickcheck_arbitrary_derive:quickcheck_arbitrary_derive",
    ],
)
