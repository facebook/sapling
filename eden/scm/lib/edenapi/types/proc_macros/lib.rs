/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

extern crate proc_macro;

use std::collections::HashSet;

use proc_macro2::TokenStream;
use quote::format_ident;
use quote::quote;
use syn::spanned::Spanned;
use syn::*;

const ID: &str = "id";
const WIRE_OPTION: &str = "wire_option";

/// Derive a default implementation for a wire type for this type
/// Supports: 'wire_option' attribute. Wire impl will wrap it in Option,
//             and fail when deserializing if it's not present.
// This is useful fo safe migration off Option types in Api types.
//
// TODO: Future improvements
// - Support fields that do not implement Default on Api obj
//    - add "no_default" attribute to field. Wire impl will wrap it in Option,
//      and fail when deserializing if it's not present
// - Support generics in type
//    - Might be possible to make it work with some adaptation, at least for
//      simple "wrapper" types that just have one generic T.
#[proc_macro_attribute]
pub fn auto_wire(
    attr: proc_macro::TokenStream,
    item: proc_macro::TokenStream,
) -> proc_macro::TokenStream {
    let _ = parse_macro_input!(attr as syn::parse::Nothing);
    let mut item = parse_macro_input!(item as Item);

    let result = match &mut item {
        Item::Struct(item) => get_wire_struct(item),
        Item::Enum(item) => get_wire_enum(item),
        item => Err(Error::new(item.span(), "Only struct or enum is supported")),
    };
    match result {
        Ok(wire_item) => quote! {
            #item

            #wire_item
        },
        Err(e) => e.to_compile_error(),
    }
    .into()
}

struct Parenthesized<T>(T);
impl<T> parse::Parse for Parenthesized<T>
where
    T: std::str::FromStr,
    T::Err: std::fmt::Display,
{
    fn parse(input: parse::ParseStream) -> Result<Self> {
        let content;
        parenthesized!(content in input);
        Ok(Self(content.parse::<LitInt>()?.base10_parse::<T>()?))
    }
}

fn arbitrary_impl(wire_ident: &Ident, generics: &Generics) -> TokenStream {
    quote! {
        #[cfg(any(test, feature = "for-tests"))]
        impl #generics quickcheck::Arbitrary for #wire_ident #generics {
            fn arbitrary(g: &mut quickcheck::Gen) -> Self {
                use crate::ToWire;
                <Self as crate::ToApi>::Api::arbitrary(g).to_wire()
            }
        }
    }
}

fn extract_id(
    attrs: Vec<Attribute>,
    spanned: &impl Spanned,
    ids: &mut HashSet<u16>,
) -> Result<(u16, Vec<Attribute>)> {
    let (id, other_attrs): (Vec<_>, Vec<_>) =
        attrs.into_iter().partition(|attr| attr.path.is_ident(ID));
    if id.len() != 1 {
        return Err(Error::new(
            spanned.span(),
            "Must have exactly one attribute 'id'",
        ));
    }
    let id = id
        .into_iter()
        .next()
        // never panics because of if above
        .unwrap();
    let id = parse2::<Parenthesized<u16>>(id.tokens)?.0;
    if !ids.insert(id) {
        return Err(Error::new(
            spanned.span(),
            "'id' attribute should be unique across all fields/variants",
        ));
    }
    Ok((id, other_attrs))
}

fn remove_id(attrs: &mut Vec<Attribute>) {
    *attrs = std::mem::take(attrs)
        .into_iter()
        .filter(|attr| !attr.path.is_ident(ID))
        .collect();
}

fn extract_no_default(attrs: &mut Vec<Attribute>) -> bool {
    let mut no_default = false;
    attrs.retain(|a| match a.path.get_ident() {
        Some(id) if *id == "no_default" => {
            no_default = true;
            false
        }
        _ => true,
    });
    no_default
}

fn extract_wire_option(attrs: &mut Vec<Attribute>) -> bool {
    let mut wire_option = false;
    attrs.retain(|a| match a.path.get_ident() {
        Some(id) if *id == WIRE_OPTION => {
            wire_option = true;
            false
        }
        _ => true,
    });
    wire_option
}

fn get_wire_struct(original: &mut ItemStruct) -> Result<TokenStream> {
    let mut item = original.clone();
    let ident = item.ident.clone();
    let wire_ident = format_ident!("Wire{}", ident);
    item.ident = wire_ident.clone();

    let mut fields = vec![];
    let mut has_no_default_field = false;
    let mut wire_option_fields = HashSet::new();

    let mut ids = HashSet::new();
    match &mut item.fields {
        Fields::Named(fs) => fs.named.iter_mut().try_for_each(|ref mut field| {
            let name = field.ident.clone().unwrap();
            fields.push(name.clone());
            let (id, other_attrs) = extract_id(std::mem::take(&mut field.attrs), &field, &mut ids)?;
            field.attrs = other_attrs;
            let ty = &field.ty;
            if extract_wire_option(&mut field.attrs) {
                field.ty = parse_quote!( <Option<#ty> as crate::ToWire>::Wire );
                wire_option_fields.insert(name);
            }
            else {
                field.ty = parse_quote!( <#ty as crate::ToWire>::Wire );
            }
            let name = format!("{}", id);

            if extract_no_default(&mut field.attrs) {
                has_no_default_field = true;
                field.attrs.push(
                    parse_quote!( #[serde(rename=#name)] ),
                );
            } else {
                field.attrs.push(
                    parse_quote!( #[serde(rename=#name, default, skip_serializing_if="crate::wire::is_default")] ),
                );
            }
            Result::Ok(())
        })?,
        _ => {
            return Err(Error::new(
                item.fields.span(),
                "Only structs with named fields supported",
            ));
        }
    }

    if has_no_default_field {
        item.attrs = vec![
            parse_quote!(#[derive(Debug, serde::Serialize, serde::Deserialize, Clone, PartialEq, Eq)]),
        ];
    } else {
        item.attrs = vec![
            parse_quote!(#[derive(Default, Debug, serde::Serialize, serde::Deserialize, Clone, PartialEq, Eq)]),
        ];
    }

    // remove id() attribute from original struct
    match &mut original.fields {
        Fields::Named(fs) => fs.named.iter_mut().for_each(|ref mut field| {
            remove_id(&mut field.attrs);
            extract_no_default(&mut field.attrs);
            extract_wire_option(&mut field.attrs);
        }),
        _ => unreachable!(),
    }

    let fields_to_wire = fields.iter().map(|name| {
        if wire_option_fields.contains(name) {
            quote! { #name: Some(self.#name.to_wire()) }
        } else {
            quote! { #name: self.#name.to_wire() }
        }
    });

    let generics = &original.generics;

    let to_wire_impl = quote! {
        impl #generics crate::ToWire for #ident #generics {
            type Wire = #wire_ident #generics;

            fn to_wire(self) -> Self::Wire {
                Self::Wire {
                    #( #fields_to_wire ),*
                }
            }
        }
    };

    let fields_to_api = fields.iter().map(|name| {
        if wire_option_fields.contains(name) {
            quote! { #name: self.#name.to_api()?.ok_or_else(|| crate::WireToApiConversionError::MissingField(stringify!(#name)))? }
        } else {
            quote! { #name: self.#name.to_api()? }
        }
    });

    let to_api_impl = quote! {
        impl #generics crate::ToApi for #wire_ident #generics {
            type Api = #ident #generics;
            type Error = crate::WireToApiConversionError;

            fn to_api(self) -> Result<Self::Api, Self::Error> {
                Ok(Self::Api {
                    #( #fields_to_api ),*
                })
            }
        }
    };

    let arbitrary_impl = arbitrary_impl(&wire_ident, generics);

    Ok(quote! {
        #item
        #to_wire_impl
        #to_api_impl
        #arbitrary_impl
    })
}

fn get_wire_enum(original: &mut ItemEnum) -> Result<TokenStream> {
    // original is the original Enum
    // item is the counterpart WireEnum
    let mut item = original.clone();
    let ident = item.ident.clone();
    let wire_ident = format_ident!("Wire{}", ident);
    item.ident = wire_ident.clone();
    item.attrs = vec![
        parse_quote!(#[derive(Debug, serde::Serialize, serde::Deserialize, Clone, PartialEq, Eq)]),
    ];

    let mut variants = vec![];
    let mut ids = HashSet::new();
    item.variants.iter_mut().try_for_each(|ref mut variant| {
        let (id, other_attrs) = extract_id(std::mem::take(&mut variant.attrs), &variant, &mut ids)?;
        variant.attrs = other_attrs;
        if id == 0 {
            return Err(Error::new(variant.span(), "Variant id can't be 0"));
        }
        let name = format!("{}", id);
        variant.attrs.push(parse_quote!( #[serde(rename=#name)]));
        let unit = match &mut variant.fields {
            Fields::Unit => true,
            Fields::Unnamed(fields) if fields.unnamed.len() == 1 => {
                let field = fields.unnamed.first_mut().unwrap();
                let ty = &field.ty;
                field.ty = parse_quote!( <#ty as crate::ToWire>::Wire );
                false
            }
            _ => {
                return Err(Error::new(
                    variant.fields.span(),
                    "Only unit variants or with a single field supported",
                ));
            }
        };
        variants.push((variant.ident.clone(), unit));
        Ok(())
    })?;
    item.variants.push_value(parse_quote!(
        #[serde(other, rename = "0")]
        UnknownVariant
    ));

    // remove id() attribute from original enum
    original.variants.iter_mut().for_each(|ref mut variant| {
        remove_id(&mut variant.attrs);
    });

    let variants_to_wire = variants.iter().map(|(name, unit)| {
        if *unit {
            quote! { Self::#name => Self::Wire::#name }
        } else {
            quote! { Self::#name(value) => Self::Wire::#name(value.to_wire()) }
        }
    });

    let generics = &original.generics;

    let to_wire_impl = quote! {
        impl #generics crate::ToWire for #ident #generics {
            type Wire = #wire_ident #generics;

            fn to_wire(self) -> Self::Wire {
                match self {
                    #( #variants_to_wire ),*
                }
            }
        }
    };

    let variants_to_api = variants.iter().map(|(name, unit)| {
        if *unit {
            quote! { Self::#name => Ok(Self::Api::#name) }
        } else {
            quote! { Self::#name(value) => Ok(Self::Api::#name(value.to_api()?)) }
        }
    });

    let to_api_impl = quote! {
        impl #generics crate::ToApi for #wire_ident #generics {
            type Api = #ident #generics;
            type Error = crate::WireToApiConversionError;

            fn to_api(self) -> Result<Self::Api, Self::Error> {
                match self {
                    #( #variants_to_api, )*
                    Self::UnknownVariant => Err(Self::Error::UnrecognizedEnumVariant(stringify!(#wire_ident))),
                }
            }
        }
    };

    let arbitrary_impl = arbitrary_impl(&wire_ident, generics);

    let default_impl = quote! {
        impl #generics Default for #wire_ident #generics {
            fn default() -> Self {
                use crate::ToWire;
                #ident::default().to_wire()
            }
        }
    };

    Ok(quote! {
        #item
        #to_wire_impl
        #to_api_impl
        #arbitrary_impl
        #default_impl
    })
}

#[test]
#[should_panic = "'id' attribute should be unique across all fields/variants"]
fn test_same_ids() {
    let mut input = parse_quote! {
        enum MyEnum {
            #[id(1)]
            A,
            #[id(1)]
            B,
        }
    };
    get_wire_enum(&mut input).unwrap();
}

#[test]
#[should_panic = "Must have exactly one attribute 'id'"]
fn test_two_ids() {
    let mut input = parse_quote! {
        enum MyEnum {
            #[id(1)]
            #[id(1)]
            A,
        }
    };
    get_wire_enum(&mut input).unwrap();
}
