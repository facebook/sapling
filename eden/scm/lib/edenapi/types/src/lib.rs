/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This software may be used and distributed according to the terms of the
 * GNU General Public License version 2.
 */

//! Types shared between the EdenAPI client and server.
//!
//! This crate exists primarily to provide a lightweight place to
//! put types that need to be used by both the client and server.
//! Types that are exclusive used by either the client or server
//! SHOULD NOT be added to this crate.
//!
//! Given that the client and server are each part of different
//! projects (Mercurial and Mononoke, respectively) which have
//! different build processes, putting shared types in their own
//! crate decreases the likelihood of build failures caused by
//! dependencies with complex or esoteric build requirements.
//!
//! Most of the types in this crate are used for data interchange
//! between the client and server. As such, CHANGES TO THE THESE
//! TYPES MAY CAUSE VERSION SKEW, so any changes should proceed
//! with caution.

#[macro_use]
pub mod hash;

pub mod anyid;
pub mod batch;
pub mod bookmark;
pub mod commit;
pub mod commitid;
pub mod errors;
pub mod file;
pub mod history;
pub mod land;
pub mod metadata;
pub mod token;
pub mod tree;
pub mod wire;

use bytes::Bytes;
// re-export CloneData
pub use dag_types::CloneData;
pub use dag_types::FlatSegment;
pub use dag_types::Location as CommitLocation;
pub use dag_types::PreparedFlatSegments;
#[cfg(any(test, feature = "for-tests"))]
use quickcheck_arbitrary_derive::Arbitrary;
use thiserror::Error;
pub use types::hgid::HgId;
pub use types::key::Key;
pub use types::nodeinfo::NodeInfo;
pub use types::parents::Parents;
pub use types::path::RepoPathBuf;

pub use crate::anyid::AnyId;
pub use crate::anyid::LookupRequest;
pub use crate::anyid::LookupResponse;
pub use crate::anyid::LookupResult;
pub use crate::batch::Batch;
pub use crate::bookmark::BookmarkEntry;
pub use crate::bookmark::BookmarkRequest;
pub use crate::bookmark::SetBookmarkRequest;
pub use crate::commit::make_hash_lookup_request;
pub use crate::commit::AlterSnapshotRequest;
pub use crate::commit::AlterSnapshotResponse;
pub use crate::commit::BonsaiChangesetContent;
pub use crate::commit::BonsaiFileChange;
pub use crate::commit::CommitGraphEntry;
pub use crate::commit::CommitGraphRequest;
pub use crate::commit::CommitHashLookupRequest;
pub use crate::commit::CommitHashLookupResponse;
pub use crate::commit::CommitHashToLocationRequestBatch;
pub use crate::commit::CommitHashToLocationResponse;
pub use crate::commit::CommitKnownResponse;
pub use crate::commit::CommitLocationToHashRequest;
pub use crate::commit::CommitLocationToHashRequestBatch;
pub use crate::commit::CommitLocationToHashResponse;
pub use crate::commit::CommitMutationsRequest;
pub use crate::commit::CommitMutationsResponse;
pub use crate::commit::CommitRevlogData;
pub use crate::commit::CommitRevlogDataRequest;
pub use crate::commit::CommitTranslateIdRequest;
pub use crate::commit::CommitTranslateIdResponse;
pub use crate::commit::EphemeralPrepareRequest;
pub use crate::commit::EphemeralPrepareResponse;
pub use crate::commit::Extra;
pub use crate::commit::FetchSnapshotRequest;
pub use crate::commit::FetchSnapshotResponse;
pub use crate::commit::HgChangesetContent;
pub use crate::commit::HgMutationEntryContent;
pub use crate::commit::SnapshotRawData;
pub use crate::commit::SnapshotRawFiles;
pub use crate::commit::UploadBonsaiChangesetRequest;
pub use crate::commit::UploadHgChangeset;
pub use crate::commit::UploadHgChangesetsRequest;
pub use crate::commit::UploadSnapshotResponse;
pub use crate::commitid::BonsaiChangesetId;
pub use crate::commitid::CommitId;
pub use crate::commitid::CommitIdScheme;
pub use crate::commitid::GitSha1;
pub use crate::errors::ServerError;
pub use crate::file::FileAttributes;
pub use crate::file::FileAuxData;
pub use crate::file::FileContent;
pub use crate::file::FileEntry;
pub use crate::file::FileError;
pub use crate::file::FileRequest;
pub use crate::file::FileResponse;
pub use crate::file::FileSpec;
pub use crate::file::HgFilenodeData;
pub use crate::file::UploadHgFilenodeRequest;
pub use crate::file::UploadTokensResponse;
pub use crate::history::HistoryEntry;
pub use crate::history::HistoryRequest;
pub use crate::history::HistoryResponse;
pub use crate::history::HistoryResponseChunk;
pub use crate::history::WireHistoryEntry;
pub use crate::land::LandStackRequest;
pub use crate::land::LandStackResponse;
pub use crate::land::PushVar;
pub use crate::metadata::AnyFileContentId;
pub use crate::metadata::ContentId;
pub use crate::metadata::DirectoryMetadata;
pub use crate::metadata::DirectoryMetadataRequest;
pub use crate::metadata::FileMetadata;
pub use crate::metadata::FileMetadataRequest;
pub use crate::metadata::FileType;
pub use crate::metadata::FsnodeId;
pub use crate::metadata::Sha1;
pub use crate::metadata::Sha256;
pub use crate::token::FileContentTokenMetadata;
pub use crate::token::IndexableId;
pub use crate::token::UploadToken;
pub use crate::token::UploadTokenData;
pub use crate::token::UploadTokenMetadata;
pub use crate::token::UploadTokenSignature;
pub use crate::tree::TreeAttributes;
pub use crate::tree::TreeChildDirectoryEntry;
pub use crate::tree::TreeChildEntry;
pub use crate::tree::TreeChildFileEntry;
pub use crate::tree::TreeEntry;
pub use crate::tree::TreeError;
pub use crate::tree::TreeRequest;
pub use crate::tree::UploadTreeEntry;
pub use crate::tree::UploadTreeRequest;
pub use crate::tree::UploadTreeResponse;
pub use crate::wire::ToApi;
pub use crate::wire::ToWire;
pub use crate::wire::WireToApiConversionError;

#[derive(Debug, Error)]
#[error("Invalid hash: {expected} (expected) != {computed} (computed)")]
pub struct InvalidHgId {
    expected: HgId,
    computed: HgId,
    data: Bytes,
    parents: Parents,
}

#[derive(Clone, Debug, Error, Eq, PartialEq)]
#[error("Error fetching key {key:?}: {err}")]
#[cfg_attr(any(test, feature = "for-tests"), derive(Arbitrary))]
pub struct EdenApiServerError {
    pub err: EdenApiServerErrorKind,
    pub key: Option<Key>,
}

impl EdenApiServerError {
    pub fn new(err: impl std::fmt::Debug) -> EdenApiServerError {
        EdenApiServerError {
            err: EdenApiServerErrorKind::OpaqueError(format!("{:?}", err)),
            key: None,
        }
    }

    pub fn with_key(key: Key, err: impl std::fmt::Debug) -> EdenApiServerError {
        EdenApiServerError {
            err: EdenApiServerErrorKind::OpaqueError(format!("{:?}", err)),
            key: Some(key),
        }
    }

    pub fn with_path(path: RepoPathBuf, err: impl std::fmt::Debug) -> EdenApiServerError {
        EdenApiServerError {
            err: EdenApiServerErrorKind::OpaqueError(format!("{:?}", err)),
            key: Some(Key {
                path,
                hgid: *HgId::null_id(),
            }),
        }
    }

    pub fn with_hgid(hgid: HgId, err: impl std::fmt::Debug) -> EdenApiServerError {
        EdenApiServerError {
            err: EdenApiServerErrorKind::OpaqueError(format!("{:?}", err)),
            key: Some(Key {
                hgid,
                path: RepoPathBuf::new(),
            }),
        }
    }
}

#[derive(Clone, Debug, Error, Eq, PartialEq)]
#[cfg_attr(any(test, feature = "for-tests"), derive(Arbitrary))]
pub enum EdenApiServerErrorKind {
    #[error("EdenAPI server returned an error with message: {0}")]
    OpaqueError(String),
}
