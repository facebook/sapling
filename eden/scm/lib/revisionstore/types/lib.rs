/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This software may be used and distributed according to the terms of the
 * GNU General Public License version 2.
 */

//! revisionstore_types - Data types used by the revisionstore crate

mod datastore;

pub use crate::datastore::Metadata;
