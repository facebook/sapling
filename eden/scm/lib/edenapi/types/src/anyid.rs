/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

use std::num::NonZeroU64;

#[cfg(any(test, feature = "for-tests"))]
use quickcheck_arbitrary_derive::Arbitrary;
use serde_derive::Deserialize;
use serde_derive::Serialize;
use type_macros::auto_wire;
use types::HgId;

use crate::AnyFileContentId;
use crate::GitSha1;
use crate::IndexableId;
use crate::UploadToken;
use crate::commitid::BonsaiChangesetId;

#[auto_wire]
#[derive(Clone, Copy, Debug, PartialEq, Eq, Serialize, Deserialize, Hash)]
#[cfg_attr(any(test, feature = "for-tests"), derive(Arbitrary))]
pub enum AnyId {
    #[id(1)]
    AnyFileContentId(AnyFileContentId),
    #[id(2)]
    HgFilenodeId(HgId),
    #[id(3)]
    HgTreeId(HgId),
    #[id(4)]
    HgChangesetId(HgId),
    #[id(5)]
    BonsaiChangesetId(BonsaiChangesetId),
    #[id(6)]
    GitChangesetId(GitSha1),
}

impl Default for AnyId {
    fn default() -> Self {
        Self::AnyFileContentId(AnyFileContentId::default())
    }
}

#[auto_wire]
#[derive(Clone, Default, Debug, Serialize, Deserialize, Eq, PartialEq)]
#[cfg_attr(any(test, feature = "for-tests"), derive(Arbitrary))]
pub struct LookupRequest {
    #[id(1)]
    pub id: AnyId,
    #[id(2)]
    pub bubble_id: Option<NonZeroU64>,
    /// If present and the original id is not, lookup will also look into this
    /// bubble, and if the id is present, copy it to the requested bubble.
    #[id(3)]
    pub copy_from_bubble_id: Option<NonZeroU64>,
}

#[auto_wire]
#[derive(Clone, Serialize, Deserialize, Debug, Eq, PartialEq)]
#[cfg_attr(any(test, feature = "for-tests"), derive(Arbitrary))]
pub enum LookupResult {
    /// Id was present, upload token for it is returned
    #[id(1)]
    Present(UploadToken),
    /// Id was not present, only its id is returned
    #[id(2)]
    NotPresent(IndexableId),
    // Possible to add an Error variant in the future if we don't want to
    // swallow the errors
}

impl Default for LookupResult {
    fn default() -> Self {
        Self::NotPresent(Default::default())
    }
}

#[auto_wire]
#[derive(Clone, Serialize, Deserialize, Debug, Eq, PartialEq)]
#[cfg_attr(any(test, feature = "for-tests"), derive(Arbitrary))]
pub struct LookupResponse {
    #[id(3)]
    pub result: LookupResult,
}
