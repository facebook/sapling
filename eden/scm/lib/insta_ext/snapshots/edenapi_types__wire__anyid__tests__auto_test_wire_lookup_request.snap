---
source: anyid.rs
expression: "WireLookupRequest :: arbitrary(& mut g)"
---
{"1":{"5":[154,77,0,155,19,212,76,218,174,127,211,79,1,180,239,159,71,129,237,47,133,207,59,183,155,157,133,85,4,46,38,85]},"3":17821718004141110471}
